fs/btrfs/compression.c

```diff
@@ -1595,12 +1606,26 @@ int btrfs_compress_heuristic(struct inode *inode, u64 start, u64 end)
 
 unsigned int btrfs_compress_str2level(const char *str)
 {
-	if (strncmp(str, "zlib", 4) != 0)
-		return 0;
+	int ret;
+	unsigned int level;
+	unsigned int default_level = 0;
+	unsigned int max_level = 0;
+
+	if (strncmp(str, "zlib", 4) == 0) {
+		default_level = BTRFS_ZLIB_DEFAULT_LEVEL;
+		max_level = BTRFS_ZLIB_MAX_LEVEL;
+	} else if (strncmp(str, "zstd", 4) == 0) {
+		default_level = BTRFS_ZSTD_DEFAULT_LEVEL;
+		max_level = BTRFS_ZSTD_MAX_LEVEL;
+	} else {
+		return default_level;
+	}
```

Comment:
Sorry, I gave you the wrong advice here, I forgot that `btrfs_compress_op` is available in this function.

```c
	int type;
 	if (strncmp(str, "zlib", 4) == 0) {
		type = BTRFS_COMPRESS_ZLIB;
	} else if (strncmp(str, "zstd", 4) == 0) {
		type = BTRFS_COMPRESS_ZSTD;
	} else {
		return default_level;
	}
	// Use btrfs_compress_op[type-1]->default_level and btrfs_compress_op[type-1]->max_level
```

Then put the macros back in their `.c` files.

Reply:
BTRFS_ZLIB_DEFAULT_LEVEL is used in super.c as well so I am not sure if I should move it into zlib.c.